Hand tele tweaks

HMBGERDO opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## What Does This PR Do
<!-- Include a small to medium description of what your PR changes. -->
<!-- Document all changes, as not doing this may delay reviews or even discourage maintainers from merging your PR! -->
<!-- If your PR fixes an issue, add "Fixes #1234" somewhere in the PR description. This will automatically close the bug upon PR submission. -->
Hand teleporter portals now have pairs, one creates next to you, second on its target turf(like TG has em)
Hand teleporter portals now doesnt have static 5% fail chance to teleport you out of the game, instead they have stable and unstable states. Whenever someone(or something) crosses the portal, it will be unstable for 2 seconds and portal will not be able to teleport anything during 0.5 seconds. During this time fail chance will be 33% and portal will be orange.
Hand teleporter now affected by EMPs. If it got EMPd, it will not be able to create portals at least for one minute(more EMP power more time). 
Hand teleporter now can disable activated portals(only hand teleporter type) by clicking them with it

## Why It's Good For The Game
<!-- Add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->
I hate random. Before changes hand tele had 5% chance to teleport in deep space and somehow i rolled all those chances. You cant play against it, you cant do anything about it you just... hope for random. I hate this. 
Currently if criminal is using hand teleporter you should just kill them before they teleport out or hope to not roll 5% chance. With that change you will be able to see when you will teleport to space and when you will teleport to normal area.
Also i like TG style paired portals allows us to teleport back if something is weird so i added this too.

I'd like to see more using of hand tele in game, but 5% chance to die isnt fun.

## Images of changes
<!-- If you did not make a map or sprite edit, you may delete this section. You may include a gif or mp4 of your feature if you want. -->
Normal portal
![image](https://github.com/ParadiseSS13/Paradise/assets/61080616/8a69a91c-576c-429e-a36b-c0a56fca331b)

Unstable portal
![image](https://github.com/ParadiseSS13/Paradise/assets/61080616/109a375f-e51d-4f5a-99a9-aef8a4c39272)

Hand tele examine on EMP
![image](https://github.com/ParadiseSS13/Paradise/assets/61080616/5d7694ba-08a5-4c29-b0c6-8a4d4f90c167)

Hand tele activation message while EMPd
![image](https://github.com/ParadiseSS13/Paradise/assets/61080616/69869289-e7ed-473c-9e18-a093bcf7c3cf)

## Testing
<!-- How did you test the PR, if at all? -->
Compiled, teleported

## Changelog
:cl:
add: Added new things
del: Removed old things
tweak: Hand teleporter now creates pair of portals instead of one
tweak: Hand teleporter portals teleport you right on portals, not around target zone
tweak: Hand teleporter portals can be deactivated by hand teleporter(by clicking them)
tweak: Hand teleporter portals now have 33% fail chance and orange icon if something or someone just teleported through them
tweak: Hand teleporter portals now has 0.5 seconds safe time after becoming unstable to let you think if you want to use it. During this time teleporting through any of paired portals is impossible.
tweak: Hand teleporter is affected by EMP. If EMPd, it will be unable to create portals for a long time(minute or more, depends on EMP severity) and will inform you about it if you examine or activate it.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->
<!-- If a PR has no impact on players (i.e. a code refactor that does not change functionality) then the entire Changelog heading and contents can be removed. -->

Qwertytoforty commented:
An interesting PR!
The EMP weakness is long overdue and a great change.  However, probably would tune it down a little, most emp disable effects are much shorter than that, perhaps 10-30 seconds based on severity.

2 way portals, combined with portal closing. Not sure on that one. It makes it much harder to space someone via teleporter sabatoge, as they can just walk back into the portal, as well as avoid ambushes. Going through a hand teleporter portal should have risk, being able to double back negates that risk greatly. If someone reconfigures both teleporters to deep space to get you, you should be stuck there, not able to walk right back through the portal after 2 seconds like nothing happened. If people are waiting for you with stunprods on the other side of the portal, you should have to use the hand teleporter again to escape, not once again walk back through the portal.

As well, normally people can follow you through a portal. Being able to close the portal behind you is going to be a bad one as well, people should be able to follow you.

Cooldown is a thing. I am not sure if it is needed. and the 0.5 second block feels a little handholdy. RNG can be bad, but it also punishes hand teleporter overuse, especially on the captain with the NAD, and unless you are not the one with the teleporter, you can always teleport back to station. I think it should stay as chance, to be honest, but I am not as sold as that as I am with the issue of 2 way portals.

HMBGERDO commented:
> 2 way portals, combined with portal closing. Not sure on that one. It makes it much harder to space someone via teleporter sabatoge, as they can just walk back into the portal, as well as avoid ambushes. Going through a hand teleporter portal should have risk, being able to double back negates that risk greatly. If someone reconfigures both teleporters to deep space to get you, you should be stuck there, not able to walk right back through the portal after 2 seconds like nothing happened. If people are waiting for you with stunprods on the other side of the portal, you should have to use the hand teleporter again to escape, not once again walk back through the portal.

If we keep two portals scheme:
1. Guys will know, portal appeared, someone will jump in
2. You will not be able to teleport away at the same moment, portal will be unstable

Though, keeping one portal scheme feels fine if we keep other changes in

> As well, normally people can follow you through a portal. Being able to close the portal behind you is going to be a bad one as well, people should be able to follow you.

Probably this is the thing im fine with removing. Closing portal behind your ass is way too powerful for sure


> Cooldown is a thing. I am not sure if it is needed. and the 0.5 second block feels a little handholdy. RNG can be bad, but it also punishes hand teleporter overuse, especially on the captain with the NAD, and unless you are not the one with the teleporter, you can always teleport back to station. I think it should stay as chance, to be honest, but I am not as sold as that as I am with the issue of 2 way portals.

This unstable allowes you to teleport in deep space if you're not careful but also removes RNG from portals which is fine. You see blue portal, it will teleport you to area you want. You see orange portal? Dont jump in. If you just got teleported and getting beaten, you probably can try to teleport away asap and with 33% fail chance will go to deep space

HMBGERDO commented:
Removed hand tele ability to disable created portals
Changed EMP effect duration from 45-60 seconds on 1 severity to 25-30
Added portal examine text when its unstable and updated portal description

Henri215 commented on code/game/objects/effects/portals.dm:
```diff
@@ -2,7 +2,7 @@
 
 /obj/effect/portal
 	name = "portal"
-	desc = "Looks unstable. Best to test it with the clown."
```
RIP clown

HMBGERDO replied:
RIP

Henri215 commented on code/game/objects/effects/portals.dm:
```diff
@@ -12,15 +12,17 @@
 	var/fail_icon = "portal1"
 
-	var/precision = TRUE // how close to the portal you will teleport. FALSE = on the portal, TRUE = adjacent
+	var/precision = FALSE // how close to the portal you will teleport. FALSE = on the portal, TRUE = adjacent
```
```suggestion
	/// How close to the portal you will teleport. FALSE = on the portal, TRUE = adjacent
	var/precision = FALSE
```

HMBGERDO replied:
updated

HMBGERDO commented:
removed dual portals :C 
added new icon state for hand teleporter when its disabled, its screen will be dark blue and blinking

HMBGERDO commented:
added custom icons for syndicate teleporter 0, 1, 2, 3, 4 charges
4 charges: 
![image](https://github.com/ParadiseSS13/Paradise/assets/61080616/a64516b4-99bd-45f8-ac3b-885bc65346f4)
3 charges:
![image](https://github.com/ParadiseSS13/Paradise/assets/61080616/a0acee6a-b359-44c8-9472-ce0e8821a49e)
2 charges:
![image](https://github.com/ParadiseSS13/Paradise/assets/61080616/d399a0ff-887a-44b1-b7e1-31be0894caba)
1 charges:
![image](https://github.com/ParadiseSS13/Paradise/assets/61080616/e8527e84-0a8d-4096-944a-cbf3f750ed50)
0 charges:
![image](https://github.com/ParadiseSS13/Paradise/assets/61080616/0a5066bf-07e1-468d-88a4-0756a24f3480)

animations working, all blinking

lewcc commented on code/game/objects/effects/portals.dm:
```diff
@@ -143,6 +145,44 @@
+/obj/effect/portal/hand_tele/examine(mob/user, infix, suffix)
+	. = ..()
+	if(unstable_time > world.time)
+		. += "<span class='warning'>[src] is shaking, looks very unstable!</span>"
```
```suggestion
		. += "<span class='warning'>[src] is shaking, it looks very unstable!</span>"
```

lewcc commented on code/game/objects/effects/portals.dm:
```diff
@@ -143,6 +145,44 @@
+	addtimer(VARSET_CALLBACK(src, inactive, FALSE), 0.5 SECONDS) // after unstable is setted you have 0.5 safe seconds to think if you want to use it
+
+/obj/effect/portal/hand_tele/proc/check_unstable()
+	if(unstable_time - world.time > 0.1 SECONDS)
```
Why not just have a boolean that gets set in `adjust_unstable` and reset in `check_unstable()`, and use that when checking if the thing is unstable, instead of checking against a time? Should avoid the race condition you're seeing here.

HMBGERDO replied:
Because `adjust_unstable()` and `check_unstable()` could be called multiple times in the same time. 
Every time portal teleports someone, it calls `adjust_unstable()` which starts 2 seconds timer to check if we should reset unstable in proc `check_unstable()`.
But, if portal is already unstable, it means timer is ticking and in two seconds it will became blue and safe and what happends if we teleport something now? This variable allows me to just add 2 seconds from last teleportation and ignore all other delayed stabilizations

lewcc replied:
Okay, I see what you mean now. I didn't realize that this could get called multiple times like that.

I'm still a little wary on this "check to see that we're just barely being called at the right time" code. It might very well be safe, but it feels like it could invite some weird race conditions. Something else that I think you could do would be to pass the current unstable time to the callback as a parameter, and then in the callback function you could check to see if the portal's `unstable_time` is greater than the value that was passed in as a parameter. If that's the case, return early.

HMBGERDO replied:
i think i did what you want in https://github.com/ParadiseSS13/Paradise/pull/21224/commits/cdc6762945f99af1f36b59e455da248c5a0f62e5

lewcc commented on code/game/objects/items/weapons/teleportation.dm:
```diff
@@ -107,14 +107,21 @@ Frequency:
 	var/turf/current_location = get_turf(user)//What turf is the user on?
+	if(emp_timer > world.time)
+		do_sparks(5, 0, loc)
+		to_chat(user, "<span class='warning'>\The [src] attempts to create portal and suddenly disables.</span>")
```
```suggestion
		to_chat(user, "<span class='warning'>[src] attempts to create portal, but abruptly shuts off.</span>")
```

lewcc commented on code/game/objects/items/weapons/teleportation.dm:
```diff
@@ -146,10 +153,28 @@ Frequency:
+	return ..()
+
+/obj/item/hand_tele/proc/emp_update_icon()
+	if(emp_timer - world.time > 0.1 SECONDS)
```
Same thing here as before. Just use a boolean like `emped` or something on the hand tele itself, and let the callback handle the time stuff instead of trying to hope you get the times lined up

lewcc commented on code/game/objects/items/weapons/teleportation.dm:
```diff
@@ -146,10 +153,28 @@ Frequency:
+/obj/item/hand_tele/examine(mob/user)
+	. = ..()
+	if(emp_timer > world.time)
+		. += "<span class='warning'>Currently looks inactive.</span>"
```
```suggestion
		. += "<span class='warning'>It looks inactive.</span>"
```

lewcc commented on code/game/objects/items/weapons/teleportation.dm:
```diff
@@ -107,14 +107,21 @@ Frequency:
 	var/turf/current_location = get_turf(user)//What turf is the user on?
+	if(emp_timer > world.time)
+		do_sparks(5, 0, loc)
+		to_chat(user, "<span class='warning'>[src] attempts to create portal, but abruptly shuts off.</span>")
```
```suggestion
		to_chat(user, "<span class='warning'>[src] attempts to create a portal, but abruptly shuts off.</span>")
```

lewcc reviewed (commented):
One last grammar nit, but beyond that the changes look good!
